qiling/os/windows/dlls/ntdll.py

```diff
@@ -452,4 +475,238 @@ def hook_wcsstr(ql: Qiling, address: int, params):
     pid = ql.os.profile["PROCESSES"].getint("csrss.exe", fallback=12345)
-    return pid
\ No newline at end of file
+    return pid
+
+# NTSYSAPI PVOID RtlPcToFileHeader(
+#   [in]  PVOID PcValue,
+#   [out] PVOID *BaseOfImage
+# );
+@winsdkapi(cc=STDCALL, params={
+    'PcValue'    : PVOID,
+    'BaseOfImage': PVOID
+})
+def hook_RtlPcToFileHeader(ql: Qiling, address: int, params):
+    pc = params["PcValue"]
+    base_of_image_ptr = params["BaseOfImage"]
+
+    containing_image = ql.loader.find_containing_image(pc)
+
+    if containing_image:
+        base_addr = containing_image.base
+    else:
+        base_addr = 0
+
+    ql.mem.write_ptr(base_of_image_ptr, base_addr)
+    return base_addr
+
+# NTSYSAPI PRUNTIME_FUNCTION RtlLookupFunctionEntry(
+#   [in]  DWORD64               ControlPc,
+#   [out] PDWORD64              ImageBase,
+#   [out] PUNWIND_HISTORY_TABLE HistoryTable
+# );
+@winsdkapi(cc=STDCALL, params={
+    'ControlPc': PVOID,
+    'ImageBase': PVOID,
+    'HistoryTable': PVOID
+})
+def hook_RtlLookupFunctionEntry(ql: Qiling, address: int, params):
+    control_pc = params["ControlPc"]
+    image_base_ptr = params["ImageBase"]
+
+    # TODO: Make use of the history table to optimize this function.
+    # Alternatively, we could add caching to the loader, seeing as the
+    # loader is responsible for lookups in the function table.
+
+    # For simplicity, we are going to ignore the history table.
+    # history_table_ptr = params["HistoryTable"]
+
+    # This function should not be getting called on x86.
+    if ql.arch.type != QL_ARCH.X8664:
+        raise QlErrorNotImplemented("RtlLookupFunctionEntry is not implemented for x86")
+
+    containing_image = ql.loader.find_containing_image(control_pc)
+
+    if containing_image:
+        base_addr = containing_image.base
+    else:
+        base_addr = 0
+
+        return 0
+    
+    # If we got a valid location to write the image base ptr,
+    # copy it there, and proceed.
+    if image_base_ptr != 0:
+        ql.mem.write_ptr(image_base_ptr, base_addr)
+
+    # Get the base address of the function table.
+    function_table_addr = base_addr + ql.loader.function_table_lookup[base_addr]
+
+    # Look up the RUNTIME_FUNCTION entry; we are interested in the index in the table
+    # so that we can compute the address.
+    runtime_function_idx, runtime_function = ql.loader.lookup_function_entry(base_addr, control_pc)
+
+    # If a suitable function entry was found,
+    # compute its address and return.
+    if runtime_function:
+        return function_table_addr + runtime_function_idx * 12    # sizeof(RUNTIME_FUNCTION)
+    
+    return 0
+
+# NTSYSAPI
+# PRUNTIME_FUNCTION
+# RtlLookupFunctionTable (
+#     IN PVOID ControlPc,
+#     OUT PVOID *ImageBase,
+#     OUT PULONG SizeOfTable
+# );
+@winsdkapi(cc=STDCALL, params={
+    'ControlPc': PVOID,
+    'ImageBase': PVOID,
+    'SizeOfTable': PVOID
+})
+def hook_RtlLookupFunctionTable(ql: Qiling, address: int, params):
```

Comment:
I have a deja-vu with this function implementation.. is it possible to avoid code duplication by refactoring them for better code sharing?